lib/SILOptimizer/Transforms/TempRValueElimination.cpp

```diff
@@ -401,11 +412,23 @@ bool TempRValueOptPass::extendAccessScopes(
       // Is this the end of an access scope of the copy-source?
       if (!aa->isNoAlias(copySrc, endAccess->getSource())) {
         assert(endAccess->getBeginAccess()->getAccessKind() ==
-                 SILAccessKind::Read &&
+                   SILAccessKind::Read &&
                "a may-write end_access should not be in the copysrc lifetime");
         endAccessToMove = endAccess;
       }
-    } else if (endAccessToMove) {
+      continue;
+    }
+    if (auto *endBorrow = dyn_cast<EndBorrowInst>(&inst)) {
+      // Handle extending a single borrow scope only
+      if (endBorrowToMove)
+        return false;
+      // If there is a non-aliasing borrow scope, it doesn't need to be extended
+      if (!aa->isNoAlias(copySrc, endBorrow->getOperand())) {
```

Comment:
I don't think it's necessary to extend an end_borrow that may-alias. We shouldn't really care about alias analysis here. We only care if the source of the copy is indirectly produced by a begin_borrow. So, it would be sufficient to check that copySrc is a transitive use of this end_borrow's begin_borrow.

I don't know if we have a utility to specifically check if an address is guarded by a borrow scope. I think we do that somewhere else and it might be nice to have. It could be done using MemAccessUtils visitors. But meanwhile, it would be sufficient just to check `AccessedStorage.compute(copySrc) == AccessedStorage.compute(endBorrow->getOperand)`

The same is probably true of the access scope check.

Reply:
Thanks. I have used the MemAccessUtils in the latest version for borrows. I'll check and follow up for access scope.

Reply:
@meg-gupta The AccessedStorage check I posted above actually makes sense for the existing end_access case. Not the end_borrow case.

Check my reasoning here, but I think the only case that you need to fix is when a nontrivial value is borrowed, then a "class" access is used to get an address, then that address is passed to a  copy_addr src.
```
%b = begin_borrow
%a = ref_element_addr %b
copy_addr %a
```

So maybe something like this:
```
auto borrowOper = beginBorrow.getOperand();
auto srcStorage = AccessedStorage.compute(copySrc);
// Relies on findReferenceRoot being consistent with AccessedStorage.compute for all reference types.
//
// TODO: Currently handles non-class borrowed values conservatively because we don't look through
// struct_extract, etc.
bool borrowedRef =
 srcStorage.isReference()
 && (!borrowOper->getType()->isClassOrClassMetatype()
        || srcStorage.getRoot() == findReferenceRoot(borrowOper))
```
That could be packaged in a utility to check if an address depends on a borrow scope. A more precise form of that utility would override the MemAccess visitors to check if the address is a transitive use of the begin_borrow, and extend the visitor to handle struct_extract.